aptos-move/move-examples/bcs-stream/sources/stream.move

```diff
@@ -0,0 +1,211 @@
+        else if (byte == 1) {
+            true
+        }
+        else {
+            abort EMALFORMED_DATA
+        }
+    }
+
+    public fun next_address(stream: &mut BCSStream): address {
+        let data = &stream.data;
+        let cur = stream.cur;
+
+        assert!(cur + 32 <= vector::length(data), EOUT_OF_BYTES);
+
+        let res = from_bcs::to_address(vector::slice(data, cur, cur + 32));
+        stream.cur = cur + 32;
+        res
+    }
+
+    public fun next_u8(stream: &mut BCSStream): u8 {
+        let data = &mut stream.data;
+        let cur = stream.cur;
+
+        assert!(cur < vector::length(data), EOUT_OF_BYTES);
+        let res = *vector::borrow(data, cur);
+        stream.cur = cur + 1;
+        res
+    }
+
+    public fun next_u16(stream: &mut BCSStream): u16 {
+        let data = &mut stream.data;
+        let cur = stream.cur;
+
+        assert!(cur + 2 <= vector::length(data), EOUT_OF_BYTES);
+        let res =
+            (*vector::borrow(data, cur) as u16) |
+            ((*vector::borrow(data, cur + 1) as u16) << 8)
+        ;
+
+        stream.cur = stream.cur + 2;
+        res
+    }
+
+    public fun next_u32(stream: &mut BCSStream): u32 {
+        let data = &mut stream.data;
+        let cur = stream.cur;
+
+        assert!(cur + 4 <= vector::length(data), EOUT_OF_BYTES);
+        let res =
+            (*vector::borrow(data, cur) as u32) |
+            ((*vector::borrow(data, cur + 1) as u32) << 8) |
+            ((*vector::borrow(data, cur + 2) as u32) << 16) |
+            ((*vector::borrow(data, cur + 3) as u32) << 24)
+        ;
+
+        stream.cur = stream.cur + 4;
+        res
+    }
+
+    public fun next_u64(stream: &mut BCSStream): u64 {
+        let data = &mut stream.data;
+        let cur = stream.cur;
+
+        assert!(cur + 8 <= vector::length(data), EOUT_OF_BYTES);
+        let res = from_bcs::to_u64(vector::slice(data, cur, cur + 8));
+        stream.cur = cur + 8;
+
+        res
+    }
+
+    public fun next_u128(stream: &mut BCSStream): u128 {
+        let data = &mut stream.data;
+        let cur = stream.cur;
+
+        assert!(cur + 16 <= vector::length(data), EOUT_OF_BYTES);
+        let res = from_bcs::to_u128(vector::slice(data, cur, cur + 16));
+        stream.cur = cur + 16;
+
+        res
+    }
+
+    public fun next_u256(stream: &mut BCSStream): u256 {
+        let data = &mut stream.data;
+        let cur = stream.cur;
+
+        assert!(cur + 32 <= vector::length(data), EOUT_OF_BYTES);
+        let res = from_bcs::to_u256(vector::slice(data, cur, cur + 32));
+        stream.cur = cur + 32;
+
+        res
+    }
+
+    public inline fun next_vector<E>(stream: &mut BCSStream, next_elem: |&mut BCSStream| E): vector<E> {
+        let len = next_length(stream);
+        let v = vector::empty();
+
+        let i = 0;
+        while (i < len) {
+            vector::push_back(&mut v, next_elem(stream));
+            i = i + 1;
+        };
+
+        v
+    }
+
+    /*
```

Comment:
It might be more gas efficient to use these implementations for `u64`, `u128`, and `u256` (see the comment I left)

Reply:
```
public fun next_u256(stream: &mut BCSStream): u256 {
    let data = &stream.data;
    let cur = stream.cur;

    assert!(cur + 32 <= vector::length(data), EOUT_OF_BYTES);
    let res =
        (*vector::borrow(data, cur) as u256) |
        ((*vector::borrow(data, cur + 1) as u256) << 8) |
        ((*vector::borrow(data, cur + 2) as u256) << 16) |
        ((*vector::borrow(data, cur + 3) as u256) << 24) |
        ((*vector::borrow(data, cur + 4) as u256) << 32) |
        ((*vector::borrow(data, cur + 5) as u256) << 40) |
        ((*vector::borrow(data, cur + 6) as u256) << 48) |
        ((*vector::borrow(data, cur + 7) as u256) << 56) |
        ((*vector::borrow(data, cur + 8) as u256) << 64) |
        ((*vector::borrow(data, cur + 9) as u256) << 72) |
        ((*vector::borrow(data, cur + 10) as u256) << 80) |
        ((*vector::borrow(data, cur + 11) as u256) << 88) |
        ((*vector::borrow(data, cur + 12) as u256) << 96) |
        ((*vector::borrow(data, cur + 13) as u256) << 104) |
        ((*vector::borrow(data, cur + 14) as u256) << 112) |
        ((*vector::borrow(data, cur + 15) as u256) << 120) |
        ((*vector::borrow(data, cur + 16) as u256) << 128) |
        ((*vector::borrow(data, cur + 17) as u256) << 136) |
        ((*vector::borrow(data, cur + 18) as u256) << 144) |
        ((*vector::borrow(data, cur + 19) as u256) << 152) |
        ((*vector::borrow(data, cur + 20) as u256) << 160) |
        ((*vector::borrow(data, cur + 21) as u256) << 168) |
        ((*vector::borrow(data, cur + 22) as u256) << 176) |
        ((*vector::borrow(data, cur + 23) as u256) << 184) |
        ((*vector::borrow(data, cur + 24) as u256) << 192) |
        ((*vector::borrow(data, cur + 25) as u256) << 200) |
        ((*vector::borrow(data, cur + 26) as u256) << 208) |
        ((*vector::borrow(data, cur + 27) as u256) << 216) |
        ((*vector::borrow(data, cur + 28) as u256) << 224) |
        ((*vector::borrow(data, cur + 29) as u256) << 232) |
        ((*vector::borrow(data, cur + 30) as u256) << 240) |
        ((*vector::borrow(data, cur + 31) as u256) << 248)
    ;

    stream.cur = stream.cur + 32;
    res
}
```

Attaching u256 function as well